### PR TITLE
Chirp generator

### PR DESCRIPTION
**Pull request by kryksyh**
Resolves: [#7528](https://github.com/audacity/audacity/issues/7528)

*(short description of the changes and the motivation to make the changes)*

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] The title of the pull request describes an issue it addresses
- [x] If changes are extensive, then there is a sequence of easily reviewable commits
- [x] Each commit's message describes its purpose and effects
- [x] There are no behavior changes unnecessary for the stated purpose of the PR

Recommended:
- [x] Each commit compiles and runs on my machine without known undesirable changes of behavior

### CODE REVIEW

File: src/effects/builtin/tonegen/ToneView.qml [hunk 1]
```diff
@@ -66,15 +60,16 @@ EffectBase {
...
-            width: 80
-            currentText: formatNumber(tone.frequency, 6)
+            Layout.preferredWidth: 80
+
+            currentText: tone.frequencyStart
 
-            validator: DoubleInputValidator {
-                decimal: 6
+            validator: DoubleValidator {
+                bottom: 1
```
Comment (saintmatthieu) [inline src/effects/builtin/tonegen/ToneView.qml:68]: Where do you get this `DoubleValidator` from ? What does the `bottom` property determine ?

File: src/effects/builtin/tonegen/toneeffect.cpp [hunk 2]
```diff
@@ -12,6 +12,10 @@ ToneEffect::ToneEffect()
     : ToneGenBase(false), GeneratorEffect(mT0, mT1)
 {}
 
+ToneEffect::ToneEffect(bool isChirp)
```
Comment (saintmatthieu) [inline src/effects/builtin/tonegen/toneeffect.cpp:15]: Why not simplify using just one ctor with bool (or enum) ?

File: src/effects/builtin/tonegen/ToneView.qml (hunk 1 above)
Comment (kryksyh) [reply src/effects/builtin/tonegen/ToneView.qml:68]: This is a standard one https://doc.qt.io/qt-6/qml-qtquick-doublevalidator.html, top and bottom define a valid range. The difference with DoubleInputValidator is that the latter adds padding zeroes we don't need.

File: src/effects/builtin/tonegen/toneeffect.cpp [hunk 3]
```diff
@@ -65,24 +71,77 @@ void ToneEffect::setWaveform(Waveform waveform)
...
+
+double ToneEffect::amplitudeEnd() const
+{
+    return mAmplitude1;
+}
+
+void ToneEffect::setAmplitudeEnd(double amplitude)
 {
-    mAmplitude0 = mAmplitude1 = amplitude;
+    mAmplitude1 = amplitude;
```
[the remark this answers is not in the document]
Comment (kryksyh) [reply src/effects/builtin/tonegen/toneeffect.cpp:121]: sure, why not

File: src/effects/builtin/tonegen/toneeffect.cpp (hunk 2 above)
Comment (kryksyh) [reply src/effects/builtin/tonegen/toneeffect.cpp:15]: It is instantiated by `BuiltinEffectsModule::Registration` which does not allow us to pass arguments to ctor.

Comment (kryksyh) [reply src/effects/builtin/tonegen/toneeffect.cpp:15]: We could use a templated class instead, though not sure if it make the code prettier.

File: src/effects/builtin/tonegen/chirpeffect.cpp [hunk 4]
```diff
@@ -8,7 +8,7 @@ using namespace au::effects;
 const ComponentInterfaceSymbol ChirpEffect::Symbol{ XO("Chirp") };
 
 ChirpEffect::ChirpEffect()
-    : ToneGenBase(true)
+    : ToneEffect(true)
```
[the remark this answers is not in the document]
Comment (kryksyh) [reply src/effects/builtin/tonegen/chirpeffect.cpp]: Absolutely

File: src/effects/builtin/tonegen/ChirpView.qml [hunk 5]
```diff
@@ -1,7 +1,167 @@
...
+    model: chirp
+
+    ToneViewModel {
+        id: chirp
+        instanceId: root.instanceId
+    }
+
+    Component.onCompleted: {
+        chirp.init()
+        timecode.currentFormatStr = chirp.durationFormat
```
Comment (saintmatthieu) [inline src/effects/builtin/tonegen/ChirpView.qml:28]: I removed this line from the other views, I guess you can here, too.

File: src/effects/builtin/tonegen/ChirpView.qml [hunk 6]
```diff
@@ -1,7 +1,167 @@
...
+            tempo: chirp.tempo
+            upperTimeSignature: chirp.upperTimeSignature
+            lowerTimeSignature: chirp.lowerTimeSignature
+            enabled: true
+
+            onValueChanged: {
+                chirp.duration = timecode.value
+            }
+
+            onCurrentFormatChanged: {
```
Comment (saintmatthieu) [inline src/effects/builtin/tonegen/ChirpView.qml:162]: You should also be able to remove this.

File: src/effects/builtin/tonegen/toneviewmodel.cpp [hunk 7]
```diff
@@ -39,53 +42,105 @@ QList<QString> ToneViewModel::waveforms() const
...
+    const ToneEffect* const te = effect();
+    if (!te) {
+        return 0.0;
+    }
+    return te->frequencyStart();
+}
+
+void ToneViewModel::prop_setFrequencyStart(double newFrequency)
+{
+    const auto wasAllowed = isApplyAllowed();
```
Comment (saintmatthieu) [inline src/effects/builtin/tonegen/toneviewmodel.cpp:113]: Shouldn't you add
```cpp
    if (!m_inited) {
        return;
    }
```
here too? (Explanation in [this comment](https://github.com/audacity/audacity/pull/7927#issue-2745281322))

**Review by saintmatthieu — APPROVED** [review]

**Pull request merged**